### PR TITLE
Update error checking to use type(of:) rather than typecasting 

### LINKED ISSUE #185: Bug in iOS 16 DevBeta1
https://github.com/google/promises/blob/46c1e6b5ac09d8f82c991061c659f67e573d425d/Sources/Promises/Promise%2BAsync.swift#L28
String can convert to NSError in iOS 16.
This most likely is a bug in iOS 16.
Hope Apple could fix this soon.

### PR DESCRIPTION
**Pull request by dmace2**
Verify error status using type(of:) is NSError.Type rather than using as? NSError

This was raised due to iOS 16 String extending Error, meaning any Promise<String> will always return as rejected even if the actual promise code was successful

fix #185

### CODE REVIEW

File: Sources/Promises/Promise.swift [hunk 1]
```diff
@@ -40,14 +40,14 @@ public final class Promise<Value> {
...
-      switch resolution {
-      case let error as NSError:
-        self.init(error)
-      case let objCPromise as ObjCPromise<AnyObject>:
-        self.init(objCPromise)
-      default:
-        self.init(ObjCPromise<AnyObject>.__resolved(with: Promise<Value>.asAnyObject(resolution)))
-      }
+      if type(of: resolution) is NSError.Type {
+          let error = resolution as! NSError
```
Comment (ykjchen) [inline Sources/Promises/Promise.swift:44]: This codebase uses 2-space indentation.

File: Sources/Promises/Promise.swift [hunk 2]
```diff
@@ -40,14 +40,14 @@ public final class Promise<Value> {
...
+          self.init(error)
+      } else if type(of: resolution) is ObjCPromise<AnyObject>.Type {
```
Comment (ykjchen) [inline Sources/Promises/Promise.swift:46]: Was this change necessary? If not prefer reverting to minimize the scope of the change.

File: Sources/Promises/Promise.swift [hunk 3]
```diff
@@ -40,14 +40,14 @@ public final class Promise<Value> {
...
       let resolution = try work()
-      switch resolution {
-      case let error as NSError:
-        self.init(error)
-      case let objCPromise as ObjCPromise<AnyObject>:
-        self.init(objCPromise)
-      default:
-        self.init(ObjCPromise<AnyObject>.__resolved(with: Promise<Value>.asAnyObject(resolution)))
-      }
+      if type(of: resolution) is NSError.Type {
```
Comment (ykjchen) [inline Sources/Promises/Promise.swift:43]: Is it possible to go back to a switch here to minimize the change?

**Comment by dmace2** [comment]
Note: This code passed all tests as well as the additional tests located below, which stemmed from the same issue and failed prior to my changes when building against iOS 16 with SafariServices.framework connected to the package

https://github.com/google/promises/issues/88#issuecomment-572834083

File: Sources/Promises/Promise+Do.swift [hunk 4]
```diff
@@ -27,7 +27,7 @@ public extension Promise {
     let objCPromise = ObjCPromise<AnyObject>.__onQueue(queue) {
       do {
         let resolution = try work()
-        return resolution as? NSError ?? Promise<Value>.asAnyObject(resolution)
+          return type(of: resolution) is NSError.Type ? resolution as! NSError : Promise<Value>.asAnyObject(resolution)
```
Comment (ykjchen) [inline Sources/Promises/Promise+Do.swift:30]: Let's use this formatting (convention for this codebase):
```
return type(of: resolution) is NSError.Type
  ? resolution as! NSError : Promise<Value>.asAnyObject(resolution)
```

Also applies to the change in Promise+Then

**Review by ykjchen — COMMENTED** [review]
Thanks for your contribution!

Can you update the unit tests to prevent future breakages of this sort?

I'm thinking that you could add the extension (assuming the tests involve String Promises):
```
extension String: Error {}
```
(per https://github.com/google/promises/issues/185#issuecomment-1152960207)

then verify that tests break without these changes, and that tests pass with these changes.

File: Sources/Promises/Promise.swift (hunk 3 above)
Comment (dmace2) [reply Sources/Promises/Promise.swift:43]: This is not possible to go back to a switch statement here due to the new type changes

File: Sources/Promises/Promise.swift (hunk 2 above)
Comment (dmace2) [reply Sources/Promises/Promise.swift:46]: This can be undone you are right

Comment (dmace2) [reply Sources/Promises/Promise.swift]: Update: I was wrong, this cannot be undone and it must stay

Comment (ykjchen) [reply Sources/Promises/Promise.swift:46]: To clarify, would this work:
```
      } else if let objCPromise = resolution as? ObjCPromise<AnyObject> {
        self.init(objCPromise)
```

This seems to be equivalent to the old code... the new code seems like it could be subtly different (I don't know enough about Swift to be sure)... so reducing the amount of change is preferred.

File: Tests/PromisesTests/Promises+StringTests.swift [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+//      log("work1: value=\(value)")
+      return Promise { value }
+    }
+    func work2(_ value: Vec2) -> Promise<Vec2> {
+//      log("work2: value=\(value)")
+      return Promise { value }
+    }
+
+    let exp = expectation(description: "testPromiseVec2")
+    let p1 = work1(Vec2(x: 42, y: 43))
```
Comment (ykjchen) [inline Tests/PromisesTests/Promises+StringTests.swift:78]: `p1` should be renamed `promise1`, etc. to match pre-existing tests.

File: Tests/PromisesTests/Promises+StringTests.swift [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    func work1(_ value: Vec2) -> Promise<Vec2> {
+//      log("work1: value=\(value)")
+      return Promise { value }
+    }
+    func work2(_ value: Vec2) -> Promise<Vec2> {
+//      log("work2: value=\(value)")
+      return Promise { value }
+    }
+
+    let exp = expectation(description: "testPromiseVec2")
```
Comment (ykjchen) [inline Tests/PromisesTests/Promises+StringTests.swift:77]: Can this be replaced with:
```
XCTAssert(waitForPromises(timeout: 10))
```
for consistency with existing tests?

File: Sources/Promises/Promise.swift [hunk 7]
```diff
@@ -40,14 +40,12 @@ public final class Promise<Value> {
...
+        let error = resolution as! NSError
         self.init(error)
-      case let objCPromise as ObjCPromise<AnyObject>:
-        self.init(objCPromise)
-      default:
-        self.init(ObjCPromise<AnyObject>.__resolved(with: Promise<Value>.asAnyObject(resolution)))
-      }
+      } else if let objCPromise = resolution as? ObjCPromise<AnyObject> {
+          self.init(objCPromise)
+      } else { self.init(ObjCPromise<AnyObject>.__resolved(with: Promise<Value>.asAnyObject(resolution))) }
```
Comment (ykjchen) [inline Sources/Promises/Promise.swift:48]: This line should be formatted like this for consistency:
```
} else {
  self.init(ObjCPromise<AnyObject>.__resolved(with: Promise<Value>.asAnyObject(resolution)))
}
```

File: Tests/PromisesTests/Promise+StringTests.swift [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    func work2(_ value: Vec2) -> Promise<Vec2> {
+      return Promise { value }
+    }
+    let promise1 = work1(Vec2(x: "42", y: "43"))
+    let promise2 = promise1.then(work2)
+
+    XCTAssert(waitForPromises(timeout: 3))
+    XCTAssertEqual(promise2.value?.x, "42")
+    XCTAssertEqual(promise1.value?.x, "42")
+
```
Comment (ykjchen) [inline Tests/PromisesTests/Promise+StringTests.swift:80]: Newline above should be omitted for consistency.

**Review by ykjchen — CHANGES_REQUESTED** [review]
Looks good overall, just some formatting/documentation nits.

Thanks again for your help here!

File: Tests/PromisesTests/Promise+StringTests.swift [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+      return Promise<String> { (_, reject) in
+        reject(Test.Error.code42)
+      }
+    }
+    
+    let promise1 = work1("42")
+    let promise2 = promise1.then(work2)
+
+    XCTAssert(waitForPromises(timeout: 10))
+    XCTAssertTrue(promise2.error == Test.Error.code42)
```
[the remark this answers is not in the document]
Comment (dmace2) [reply Tests/PromisesTests/Promise+StringTests.swift:57]: This is inline with all other tests in the promise framework and will not compile using XCTAssertEqual

**Review by ykjchen — APPROVED** [review]
Nice work @dmace2, thank you!
